[7.x] Add actingAsGuest Authentication testing helper

theomessin opened the pull request:
~~Allows for actingAs on testing to be null. This is helpful when testing applications where most routes are auth protected and you have actingAs a test user in the base TestCase but you want to test certain routes as guest. You can now simply use $this->actingAs(null) or even $this->actingAs(null, 'api').~~

Adds an `actingAsGuest()` authentication helper to undo any previous `actingAs` statements. Done by re-adding a `logout` to Guards which simply sets `$this->user` to `null` if using `GuardHelpers` and performs a full `logout` for the `sessionGuard`.

See discussion in https://github.com/laravel/framework/pull/30644

<!--
Please only send a pull request to branches which are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

Commit:
Auth guards setUser $user and actingAs $user can be null

Allows for actingAs on testing to be null. This is helpful when testing applications where most routes are auth protected and you have actingAs a test user in the base TestCase but you want to test certain routes as guest. You can now simply use $this->actingAs(null) or even $this->actingAs(null, 'api').

Commit:
Address @driesvints pull request code review comments

driesvints commented:
If we're gonna do this I'd just re-add the logout method to make the API more expressive and not rely on passing `null`.

theomessin commented:
@driesvints no worries :smiley: done

Commit:
Remove logout method from StatefulGuard

This has been added to the parent Guard contract now.

driesvints commented:
Why did you move the logout method to a different contract?

theomessin commented:
Initially I had just added it to `Guard` but then I realised that `StatefulGuard` extends `Guard` but re-implements the same method word-for-word so I removed it from there.

We need it to be in `Guard` because `actingAsGuest` accepts any guard driver and eg. the `api` driver (`TokenGuard`) is not a `StatefulGuard` but a generic `Guard`.

driesvints commented:
But this will require any non-stateful guard to implement the guard method.

theomessin commented:
Yeah, that's what we need to be able to simply run this in actingAsGuest.
```php
$this->app['auth']->guard($driver)->logout();
```
The guard for `$driver` can either be stateful or not. Maybe logout is not the right word for what we're doing (in the case of non-stateful guards). All we're trying to do is set `this->user` to `null`. If we just keep everything as is in the contracts (at least we wouldn't have any breaking changes :fireworks:) we'd have to check whether the `$this->app['auth']->guard($driver)` is stateful or not in `actingAsGuest`. In that case `actingAsGuest` would only work for stateful guards (ie. not for `api`).

theomessin commented:
So for 6.x we could actually have an actingAsGuest which only works for stateful guards. Roughly:
```php
public function actingAsGuest($driver = null)
{
    $guard = $this->app['auth']->guard($driver);

    if (is_a($guard, StatefulGuard::class) == false) {
        throw new InvalidArgumentException('actingAsGuest $driver guard must be of type StatefulGuard');
    }

    $guard->logout();

    return $this;
}
```

rbuzzibarrios commented on tests/Integration/Foundation/Testing/Concerns/InteractsWithAuthenticationTest.php:
```diff
@@ -59,23 +60,64 @@ public function testActingAsIsProperlyHandledForSessionAuth()
+            return 'This is a login page.';
+        })->middleware(['guest'])->name('login');
+
+        $user = AuthenticationTestUser::where('username', '=', 'taylorotwell')->first();
```
```suggestion
        $user = AuthenticationTestUser::where('username', 'taylorotwell')->first();
```

rbuzzibarrios replied:
```suggestion
        $user = AuthenticationTestUser::where('username', 'taylorotwell')->first();
```

rbuzzibarrios commented on tests/Integration/Foundation/Testing/Concerns/InteractsWithAuthenticationTest.php:
```diff
@@ -59,23 +60,64 @@ public function testActingAsIsProperlyHandledForSessionAuth()
             return $request->user();
         });
-
         $user = AuthenticationTestUser::where('username', '=', 'taylorotwell')->first();
```
```suggestion
         $user = AuthenticationTestUser::where('username', 'taylorotwell')->first();
```

rbuzzibarrios commented on tests/Integration/Foundation/Testing/Concerns/InteractsWithAuthenticationTest.php:
```diff
@@ -59,23 +60,64 @@ public function testActingAsIsProperlyHandledForSessionAuth()
+            return $request->user();
+        });
+
+        $user = AuthenticationTestUser::where('username', '=', 'taylorotwell')->first();
```
```suggestion
        $user = AuthenticationTestUser::where('username', 'taylorotwell')->first();
```

dramosv90 commented on src/Illuminate/Contracts/Auth/Guard.php:
```diff
@@ -47,4 +47,11 @@ public function validate(array $credentials = []);
     public function setUser(Authenticatable $user);
+
+    /**
+     * Log the user out of the application.
```
```suggestion
     * Log out of the application user.
```

GrahamCampbell replied:
👎

taylorotwell commented:
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.
